File: packages/cli-js/src/context-optimizer.js
#!/usr/bin/env node
/**
 * PAWS Context Optimizer - Handle 500K+ line refactors
 *
 * Strategies:
 * - Hierarchical chunking with CATSCAN summaries
 * - Dependency graph analysis
 * - Smart context windowing
 * - Iterative refinement with context expansion
 */

const fs = require('fs').promises;
const fsSync = require('fs');
const path = require('path');
const { Command } = require('commander');
const chalk = require('chalk');
const { glob } = require('glob');

/**
 * Represents a code module with metadata
 */
class CodeModule {
  constructor(filePath, sizeLines, imports, exports, complexityScore, summary = null) {
    this.path = filePath;
    this.sizeLines = sizeLines;
    this.imports = imports;
    this.exports = exports;
    this.complexityScore = complexityScore;
    this.summary = summary;
  }
}

/**
 * A context window for a specific task
 */
class ContextWindow {
  constructor(coreFiles, summaryFiles, totalLines, estimatedTokens) {
    this.coreFiles = coreFiles;
    this.summaryFiles = summaryFiles;  // Files to include as CATSCAN summaries only
    this.totalLines = totalLines;
    this.estimatedTokens = estimatedTokens;
  }
}

/**
 * Analyze code dependencies to build optimal context windows
 */
class DependencyAnalyzer {
  constructor(rootPath) {
    this.rootPath = rootPath;
    this.modules = new Map();
  }

  /**
   * Analyze a JavaScript/TypeScript file
   */
  async analyzeJavaScriptFile(filePath) {
    try {
      const content = await fs.readFile(filePath, 'utf-8');
      const lines = content.split('\n').length;

      // Simple parsing for imports and exports
      const imports = new Set();
      const exports = new Set();

      // Match import statements
      const importRegex = /import\s+.*\s+from\s+['"]([^'"]+)['"]/g;
      let match;
      while ((match = importRegex.exec(content)) !== null) {
        imports.add(match[1]);
      }

      // Match require statements
      const requireRegex = /require\(['"]([^'"]+)['"]\)/g;
      while ((match = requireRegex.exec(content)) !== null) {
        imports.add(match[1]);
      }

      // Match exports
      const exportRegex = /export\s+(const|let|var|function|class|default)\s+(\w+)/g;
      while ((match = exportRegex.exec(content)) !== null) {
        exports.add(match[2]);
      }

      // Match module.exports
      const moduleExportsRegex = /module\.exports\s*=\s*\{([^}]+)\}/;
      const moduleExportsMatch = content.match(moduleExportsRegex);
      if (moduleExportsMatch) {
        const exportNames = moduleExportsMatch[1].split(',').map(e => e.trim().split(':')[0]);
        exportNames.forEach(name => exports.add(name));
      }

      // Simple complexity score based on control structures
      const complexityIndicators = [
        /\bif\b/g,
        /\bfor\b/g,
        /\bwhile\b/g,
        /\bswitch\b/g,
        /\btry\b/g,
        /\bcatch\b/g,
        /\basync\b/g,
        /\bawait\b/g
      ];

      let complexity = 0;
      complexityIndicators.forEach(regex => {
        const matches = content.match(regex);
        complexity += matches ? matches.length : 0;
      });

      const complexityScore = complexity / 100.0;

      return new CodeModule(
        filePath,
        lines,
        imports,
        exports,
        complexityScore
      );

    } catch (error) {
      console.log(chalk.yellow(`Warning: Could not analyze ${filePath}: ${error.message}`));
      return new CodeModule(
        filePath,
        0,
        new Set(),
        new Set(),
        0.0
      );
    }
  }

  /**
   * Build a dependency graph from files
   */
  async buildDependencyGraph(files) {
    const graph = new Map();

    for (const filePath of files) {
      const ext = path.extname(filePath);
      if (['.js', '.ts', '.jsx', '.tsx'].includes(ext)) {
        const module = await this.analyzeJavaScriptFile(filePath);
        this.modules.set(filePath, module);

        // Build graph edges
        const deps = new Set();

        for (const imp of module.imports) {
          // Try to resolve import to file path
          for (const otherPath of files) {
            if (imp.includes(path.basename(otherPath, path.extname(otherPath)))) {
              deps.add(otherPath);
            }
          }
        }

        graph.set(filePath, deps);
      }
    }

    return graph;
  }

  /**
   * Find transitive dependencies up to max_depth
   */
  findDependencies(filePath, graph, maxDepth = 2) {
    const dependencies = new Set();
    const toVisit = [[filePath, 0]];
    const visited = new Set();

    while (toVisit.length > 0) {
      const [current, depth] = toVisit.shift();

      if (visited.has(current) || depth > maxDepth) {
        continue;
      }

      visited.add(current);
      dependencies.add(current);

      if (graph.has(current)) {
        for (const dep of graph.get(current)) {
          if (!visited.has(dep)) {
            toVisit.push([dep, depth + 1]);
          }
        }
      }
    }

    return dependencies;
  }
}

/**
 * Optimize context windows for large codebases
 */
class ContextOptimizer {
  constructor(rootPath, maxTokens = 100000) {
    this.rootPath = rootPath;
    this.maxTokens = maxTokens;
    this.maxLines = Math.floor(maxTokens / 4);  // Rough estimate: 4 tokens per line
    this.analyzer = new DependencyAnalyzer(rootPath);
  }

  /**
   * Generate a CATSCAN summary for a file
   */
  generateCatscanSummary(filePath) {
    const module = this.analyzer.modules.get(filePath);

    if (!module) {
      return `# ${path.basename(filePath)}\n\nFile not analyzed.`;
    }

    const summaryLines = [
      `# ${path.basename(filePath)}`,
      '',
      `**Size:** ${module.sizeLines} lines`,
      `**Complexity:** ${module.complexityScore.toFixed(2)}`,
      '',
      '## Public API',
      ''
    ];

    for (const exp of Array.from(module.exports).sort()) {
      summaryLines.push(`- \`${exp}\``);
    }

    summaryLines.push('');
    summaryLines.push('## Dependencies');
    summaryLines.push('');

    for (const imp of Array.from(module.imports).sort()) {
      summaryLines.push(`- \`${imp}\``);
    }

    return summaryLines.join('\n');
  }

  /**
   * Create an optimized context window for a task
   */
  async createContextWindow(task, relevantFiles) {
    console.log(chalk.cyan(`\n☉ Optimizing context window...`));
    console.log(`Analyzing ${relevantFiles.length} files...`);

    // Build dependency graph
    const graph = await this.analyzer.buildDependencyGraph(relevantFiles);

    // Rank files by relevance
    const rankedFiles = this._rankFilesByRelevance(task, relevantFiles, graph);

    // Allocate to core vs summary
    const coreFiles = [];
    const summaryFiles = [];
    let totalLines = 0;

    for (const filePath of rankedFiles) {
      const module = this.analyzer.modules.get(filePath);
      if (!module) {
        continue;
      }

      // If adding this file keeps us under limit, add to core
      if (totalLines + module.sizeLines < this.maxLines * 0.7) {  // Reserve 30% for summaries
        coreFiles.push(filePath);
        totalLines += module.sizeLines;
      } else {
        summaryFiles.push(filePath);
      }
    }

    // Estimate tokens
    const estimatedTokens = totalLines * 4 + summaryFiles.length * 100;  // Summaries are small

    console.log(chalk.green(`\n☉ Context window created:`));
    console.log(`  Core files (full content): ${coreFiles.length}`);
    console.log(`  Summary files (CATSCAN): ${summaryFiles.length}`);
    console.log(`  Total lines: ${totalLines}`);
    console.log(`  Estimated tokens: ${estimatedTokens}`);

    return new ContextWindow(
      coreFiles,
      summaryFiles,
      totalLines,
      estimatedTokens
    );
  }

  /**
   * Rank files by relevance to task
   */
  _rankFilesByRelevance(task, files, graph) {
    // Simple heuristic: keyword matching + dependency centrality
    const taskKeywords = new Set(task.toLowerCase().split(/\s+/));

    const scores = new Map();

    for (const filePath of files) {
      let score = 0.0;

      // Keyword matching
      const fileNameLower = filePath.toLowerCase();
      for (const keyword of taskKeywords) {
        if (fileNameLower.includes(keyword)) {
          score += 10.0;
        }
      }

      // Dependency centrality (how many files depend on this one)
      let dependents = 0;
      for (const deps of graph.values()) {
        if (deps.has(filePath)) {
          dependents++;
        }
      }
      score += dependents * 2.0;

      // Module complexity (prefer simpler files in core)
      const module = this.analyzer.modules.get(filePath);
      if (module) {
        score -= module.complexityScore * 0.5;
      }

      scores.set(filePath, score);
    }

    return Array.from(files).sort((a, b) => (scores.get(b) || 0) - (scores.get(a) || 0));
  }

  /**
   * Create an optimized CATS bundle with hierarchical content
   */
  async createOptimizedBundle(window, outputPath) {
    const lines = [];

    lines.push('# Optimized Context Bundle');
    lines.push('# Generated by PAWS Context Optimizer');
    lines.push('');
    lines.push('## Context Strategy');
    lines.push(`- Core files (full): ${window.coreFiles.length}`);
    lines.push(`- Summary files: ${window.summaryFiles.length}`);
    lines.push(`- Total estimated tokens: ${window.estimatedTokens}`);
    lines.push('');

    // Core files with full content
    lines.push('## Core Files (Full Content)');
    lines.push('');

    for (const filePath of window.coreFiles) {
      try {
        const content = await fs.readFile(filePath, 'utf-8');
        const relPath = path.relative(this.rootPath, filePath);
        const ext = path.extname(filePath).substring(1);

        lines.push(`🐈 --- CATS_START_FILE: ${relPath} ---`);
        lines.push(`\`\`\`${ext}`);
        lines.push(content);
        lines.push('```');
        lines.push(`🐈 --- CATS_END_FILE: ${relPath} ---`);
        lines.push('');

      } catch (error) {
        console.log(chalk.yellow(`Warning: Could not read ${filePath}: ${error.message}`));
      }
    }

    // Summary files
    if (window.summaryFiles.length > 0) {
      lines.push('## Related Files (CATSCAN Summaries)');
      lines.push('');
      lines.push('The following files are relevant but summarized to save tokens:');
      lines.push('');

      for (const filePath of window.summaryFiles) {
        const relPath = path.relative(this.rootPath, filePath);
        const summary = this.generateCatscanSummary(filePath);
        lines.push(`### ${relPath}`);
        lines.push(summary);
        lines.push('');
      }
    }

    // Write bundle
    await fs.writeFile(outputPath, lines.join('\n'));

    console.log(chalk.green(`\n☉ Optimized bundle written to: ${outputPath}`));
  }
}

/**
 * Main CLI function
 */
async function main() {
  const program = new Command();

  program
    .name('paws-context-optimizer')
    .description('PAWS Context Optimizer - Handle massive codebases')
    .argument('[task]', 'Task description for context optimization')
    .option('--files <files...>', 'Files to include (or use --scan)')
    .option('--scan <directory>', 'Scan directory for relevant files')
    .option('--output <path>', 'Output bundle path', 'optimized_context.md')
    .option('--max-tokens <number>', 'Maximum tokens for context window', '100000')
    .parse(process.argv);

  const options = program.opts();
  let [task] = program.args;

  // Interactive prompts
  const readline = require('readline');
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  const question = (prompt) => new Promise(resolve => rl.question(prompt, resolve));

  if (!task) {
    task = await question('Enter task description:\n> ');
  }

  rl.close();

  // Determine files
  let files;
  if (options.files) {
    files = options.files;
  } else if (options.scan) {
    const scanPath = options.scan;
    console.log(chalk.gray(`Scanning ${scanPath}...`));

    const jsFiles = await glob(`${scanPath}/**/*.js`);
    const tsFiles = await glob(`${scanPath}/**/*.ts`);
    const jsxFiles = await glob(`${scanPath}/**/*.jsx`);
    const tsxFiles = await glob(`${scanPath}/**/*.tsx`);

    files = [...jsFiles, ...tsFiles, ...jsxFiles, ...tsxFiles];
    console.log(chalk.green(`Scanned ${files.length} files from ${scanPath}`));
  } else {
    console.error(chalk.red('Error: Either --files or --scan required'));
    process.exit(1);
  }

  // Create optimizer
  const rootPath = process.cwd();
  const optimizer = new ContextOptimizer(rootPath, parseInt(options.maxTokens));

  // Create optimized window
  const window = await optimizer.createContextWindow(task, files);

  // Generate bundle
  const outputPath = options.output;
  await optimizer.createOptimizedBundle(window, outputPath);

  console.log(chalk.green(`\n☉ Context optimization complete!`));
  console.log('Use this optimized bundle with:');
  console.log(chalk.cyan(`  node packages/cli-js/src/paxos.js "${task}" ${outputPath} --verify-cmd "npm test"`));

  process.exit(0);
}

// Export for use as a module
module.exports = {
  CodeModule,
  ContextWindow,
  DependencyAnalyzer,
  ContextOptimizer
};

// Run if called directly
if (require.main === module) {
  main().catch(error => {
    console.error(chalk.red('Fatal error:'), error);
    process.exit(1);
  });
}

File: packages/cli-py/paws/context_optimizer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
PAWS Context Optimizer - Handle 500K+ line refactors

Strategies:
- Hierarchical chunking with CATSCAN summaries
- Dependency graph analysis
- Smart context windowing
- Iterative refinement with context expansion
"""

import argparse
import json
import sys
import ast
from pathlib import Path
from dataclasses import dataclass
from typing import List, Dict, Set, Optional, Any
import subprocess


@dataclass
class CodeModule:
    """Represents a code module with metadata"""
    path: Path
    size_lines: int
    imports: Set[str]
    exports: Set[str]  # Public API
    complexity_score: float
    summary: Optional[str] = None


@dataclass
class ContextWindow:
    """A context window for a specific task"""
    core_files: List[Path]
    summary_files: List[Path]  # Files to include as CATSCAN summaries only
    total_lines: int
    estimated_tokens: int


class DependencyAnalyzer:
    """Analyze code dependencies to build optimal context windows"""

    def __init__(self, root_path: Path):
        self.root_path = root_path
        self.modules: Dict[str, CodeModule] = {}

    def analyze_python_file(self, file_path: Path) -> CodeModule:
        """Analyze a Python file"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
                lines = content.count('\n') + 1

            tree = ast.parse(content, filename=str(file_path))

            imports = set()
            exports = set()

            for node in ast.walk(tree):
                if isinstance(node, ast.Import):
                    for alias in node.names:
                        imports.add(alias.name)
                elif isinstance(node, ast.ImportFrom):
                    if node.module:
                        imports.add(node.module)
                elif isinstance(node, (ast.FunctionDef, ast.ClassDef)):
                    if not node.name.startswith('_'):
                        exports.add(node.name)

            # Simple complexity score based on AST depth
            complexity = len(list(ast.walk(tree))) / 100.0

            return CodeModule(
                path=file_path,
                size_lines=lines,
                imports=imports,
                exports=exports,
                complexity_score=complexity
            )

        except Exception as e:
            print(f"Warning: Could not analyze {file_path}: {e}")
            return CodeModule(
                path=file_path,
                size_lines=0,
                imports=set(),
                exports=set(),
                complexity_score=0.0
            )

    def build_dependency_graph(self, files: List[Path]) -> Dict[str, Set[str]]:
        """Build a dependency graph from files"""
        graph = {}

        for file_path in files:
            if file_path.suffix == '.py':
                module = self.analyze_python_file(file_path)
                self.modules[str(file_path)] = module

                # Build graph edges
                module_name = str(file_path)
                graph[module_name] = set()

                for imp in module.imports:
                    # Try to resolve import to file path
                    for other_path in files:
                        if imp in str(other_path):
                            graph[module_name].add(str(other_path))

        return graph

    def find_dependencies(self, file_path: Path, graph: Dict[str, Set[str]],
                         max_depth: int = 2) -> Set[str]:
        """Find transitive dependencies up to max_depth"""
        dependencies = set()
        to_visit = [(str(file_path), 0)]
        visited = set()

        while to_visit:
            current, depth = to_visit.pop(0)

            if current in visited or depth > max_depth:
                continue

            visited.add(current)
            dependencies.add(current)

            if current in graph:
                for dep in graph[current]:
                    if dep not in visited:
                        to_visit.append((dep, depth + 1))

        return dependencies


class ContextOptimizer:
    """Optimize context windows for large codebases"""

    def __init__(self, root_path: Path, max_tokens: int = 100000):
        self.root_path = root_path
        self.max_tokens = max_tokens
        self.max_lines = max_tokens // 4  # Rough estimate: 4 tokens per line
        self.analyzer = DependencyAnalyzer(root_path)

    def generate_catscan_summary(self, file_path: Path) -> str:
        """Generate a CATSCAN summary for a file"""
        module = self.analyzer.modules.get(str(file_path))

        if not module:
            return f"# {file_path.name}\n\nFile not analyzed."

        summary_lines = [
            f"# {file_path.name}",
            f"",
            f"**Size:** {module.size_lines} lines",
            f"**Complexity:** {module.complexity_score:.2f}",
            f"",
            f"## Public API",
            ""
        ]

        for export in sorted(module.exports):
            summary_lines.append(f"- `{export}`")

        summary_lines.extend([
            "",
            "## Dependencies",
            ""
        ])

        for imp in sorted(module.imports):
            summary_lines.append(f"- `{imp}`")

        return "\n".join(summary_lines)

    def create_context_window(self, task: str, relevant_files: List[Path]) -> ContextWindow:
        """Create an optimized context window for a task"""
        print(f"\n☉ Optimizing context window...")
        print(f"Analyzing {len(relevant_files)} files...")

        # Build dependency graph
        graph = self.analyzer.build_dependency_graph(relevant_files)

        # Rank files by relevance
        ranked_files = self._rank_files_by_relevance(task, relevant_files, graph)

        # Allocate to core vs summary
        core_files = []
        summary_files = []
        total_lines = 0

        for file_path in ranked_files:
            module = self.analyzer.modules.get(str(file_path))
            if not module:
                continue

            # If adding this file keeps us under limit, add to core
            if total_lines + module.size_lines < self.max_lines * 0.7:  # Reserve 30% for summaries
                core_files.append(file_path)
                total_lines += module.size_lines
            else:
                summary_files.append(file_path)

        # Estimate tokens
        estimated_tokens = total_lines * 4 + len(summary_files) * 100  # Summaries are small

        print(f"\n☉ Context window created:")
        print(f"  Core files (full content): {len(core_files)}")
        print(f"  Summary files (CATSCAN): {len(summary_files)}")
        print(f"  Total lines: {total_lines}")
        print(f"  Estimated tokens: {estimated_tokens}")

        return ContextWindow(
            core_files=core_files,
            summary_files=summary_files,
            total_lines=total_lines,
            estimated_tokens=estimated_tokens
        )

    def _rank_files_by_relevance(self, task: str, files: List[Path],
                                 graph: Dict[str, Set[str]]) -> List[Path]:
        """Rank files by relevance to task"""
        # Simple heuristic: keyword matching + dependency centrality
        task_keywords = set(task.lower().split())

        scores = {}
        for file_path in files:
            score = 0.0

            # Keyword matching
            file_name_lower = str(file_path).lower()
            for keyword in task_keywords:
                if keyword in file_name_lower:
                    score += 10.0

            # Dependency centrality (how many files depend on this one)
            dependents = sum(1 for deps in graph.values() if str(file_path) in deps)
            score += dependents * 2.0

            # Module complexity (prefer simpler files in core)
            module = self.analyzer.modules.get(str(file_path))
            if module:
                score -= module.complexity_score * 0.5

            scores[file_path] = score

        return sorted(files, key=lambda f: scores.get(f, 0), reverse=True)

    def create_optimized_bundle(self, window: ContextWindow, output_path: Path):
        """Create an optimized CATS bundle with hierarchical content"""
        lines = []

        lines.append("# Optimized Context Bundle")
        lines.append("# Generated by PAWS Context Optimizer")
        lines.append("")
        lines.append(f"## Context Strategy")
        lines.append(f"- Core files (full): {len(window.core_files)}")
        lines.append(f"- Summary files: {len(window.summary_files)}")
        lines.append(f"- Total estimated tokens: {window.estimated_tokens}")
        lines.append("")

        # Core files with full content
        lines.append("## Core Files (Full Content)")
        lines.append("")

        for file_path in window.core_files:
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()

                rel_path = file_path.relative_to(self.root_path)
                lines.append(f"🐈 --- CATS_START_FILE: {rel_path} ---")
                lines.append(f"```{file_path.suffix[1:]}")
                lines.append(content)
                lines.append("```")
                lines.append(f"🐈 --- CATS_END_FILE: {rel_path} ---")
                lines.append("")

            except Exception as e:
                print(f"Warning: Could not read {file_path}: {e}")

        # Summary files
        if window.summary_files:
            lines.append("## Related Files (CATSCAN Summaries)")
            lines.append("")
            lines.append("The following files are relevant but summarized to save tokens:")
            lines.append("")

            for file_path in window.summary_files:
                rel_path = file_path.relative_to(self.root_path)
                summary = self.generate_catscan_summary(file_path)
                lines.append(f"### {rel_path}")
                lines.append(summary)
                lines.append("")

        # Write bundle
        with open(output_path, 'w', encoding='utf-8') as f:
            f.write("\n".join(lines))

        print(f"\n☉ Optimized bundle written to: {output_path}")


def main():
    parser = argparse.ArgumentParser(
        description="PAWS Context Optimizer - Handle massive codebases"
    )

    parser.add_argument("task", nargs='?', help="Task description for context optimization")
    parser.add_argument("--files", nargs='+', help="Files to include (or use --scan)")
    parser.add_argument("--scan", help="Scan directory for relevant files")
    parser.add_argument("--output", default="optimized_context.md", help="Output bundle path")
    parser.add_argument("--max-tokens", type=int, default=100000, help="Maximum tokens for context window")

    args = parser.parse_args()

    # Interactive prompts
    task = args.task or input("Enter task description:\n> ")

    # Determine files
    if args.files:
        files = [Path(f) for f in args.files]
    elif args.scan:
        scan_path = Path(args.scan)
        files = list(scan_path.rglob("*.py")) + list(scan_path.rglob("*.js"))
        print(f"Scanned {len(files)} files from {scan_path}")
    else:
        print("Error: Either --files or --scan required")
        return 1

    # Create optimizer
    root_path = Path.cwd()
    optimizer = ContextOptimizer(root_path, max_tokens=args.max_tokens)

    # Create optimized window
    window = optimizer.create_context_window(task, files)

    # Generate bundle
    output_path = Path(args.output)
    optimizer.create_optimized_bundle(window, output_path)

    print(f"\n☉ Context optimization complete!")
    print(f"Use this optimized bundle with:")
    print(f"  paws-paxos \"{task}\" {output_path} --verify-cmd \"pytest\"")

    return 0


if __name__ == "__main__":
    sys.exit(main())
